presto-native-execution/presto_cpp/main/expression/RowExpressionEvaluator.cpp

```diff
@@ -0,0 +1,419 @@
+  cexpr.type = constantExpr->type()->toString();
+  cexpr.valueBlock.data = serializeValueBlock(constantExpr->value());
+  return std::make_shared<protocol::ConstantExpression>(cexpr);
+}
+
+json RowExpressionEvaluator::getSpecialFormRowConstructor(
+    const std::shared_ptr<const velox::exec::Expr>& expr,
+    const json& inputRowExpr) {
+  json res;
+  res["@type"] = "special";
+  res["form"] = "ROW_CONSTRUCTOR";
+  auto type = asRowType(expr->type());
+  std::string returnType = "row(";
+  for (const auto& childType : type->children()) {
+    auto typeStr = childType->toString();
+    boost::algorithm::to_lower(typeStr);
+    returnType += typeStr;
+  }
+  returnType += ")";
+  res["returnType"] = returnType;
+
+  res["arguments"] = json::array();
+  auto inputs = expr->inputs();
+  if (inputs.empty()) {
+    if (auto constantExpr =
+            std::dynamic_pointer_cast<const exec::ConstantExpr>(expr)) {
+      protocol::ConstantExpression cexpr;
+      auto typeStr = constantExpr->type()->childAt(0)->toString();
+      boost::algorithm::to_lower(typeStr);
+      cexpr.type = typeStr;
+      cexpr.valueBlock.data = serializeValueBlock(constantExpr->value());
+      json j;
+      protocol::to_json(j, cexpr);
+      res["arguments"].push_back(cexpr);
+    }
+  } else {
+    for (const auto& input : inputs) {
+      res["arguments"].push_back(veloxExprToRowExpr(input, inputRowExpr));
+    }
+  }
+
+  if (inputRowExpr.contains("sourceLocation")) {
+    res["sourceLocation"] = inputRowExpr["sourceLocation"];
+  }
+  return res;
+}
+
+json RowExpressionEvaluator::getSpecialForm(
+    const std::shared_ptr<const velox::exec::Expr>& expr,
+    const json& inputRowExpr) {
+  json res;
+  res["@type"] = "special";
+  std::string form;
+  if (inputRowExpr.contains("form")) {
+    form = inputRowExpr["form"];
+  } else {
+    // is_null is a special form expression in Presto but a call expression in
+    // Velox.
+    form = expr->name();
+  }
+  // Presto requires the field form to be in upper case.
+  std::transform(form.begin(), form.end(), form.begin(), ::toupper);
+  res["form"] = form;
+
+  auto inputs = expr->inputs();
+  res["arguments"] = json::array();
+  for (const auto& input : inputs) {
+    res["arguments"].push_back(veloxExprToRowExpr(input, inputRowExpr));
+  }
+  res["returnType"] = expr->type()->toString();
+  if (inputRowExpr.contains("sourceLocation")) {
+    res["sourceLocation"] = inputRowExpr["sourceLocation"];
+  }
+  return res;
+}
+
+// Optimizes special form expressions. Optimization rules borrowed from
+// Presto function visitSpecialForm, in
+// presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionInterpreter.java.
+std::shared_ptr<protocol::RowExpression>
+RowExpressionEvaluator::optimizeSpecialForm(
+    const std::shared_ptr<protocol::SpecialFormExpression>& specialFormExpr) {
+  switch (specialFormExpr->form) {
+    case protocol::Form::IF: {
+      auto condition = specialFormExpr->arguments[0];
```

Comment:
Make individual functions for each case value. That would make this function more readable.